In [1]:
# importing libraries

import pandas as pd
import numpy as np
import os
import pickle

In [53]:
# folder path into a string 
path = r'C:\\Users\\Jalon\\Desktop\\python_projects\\4.3_orders_products'

In [3]:
#reading the orders_products_merged data set and assigning it to a new variable
ords_prods_merged = pd.read_pickle(os.path.join(path,'orders_products_merged.pkl'))

In [5]:
ords_prods_merged.head(5)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered _merge product_name  aisle_id  department_id  prices  
0          0   both         Soda        77              7     9.0  
1          1   both         Soda        77              7     9.0  
2          1   both         Soda        77              7     9.0  
3          1   both         Soda        77              7     9.0  
4          1   both         Soda        77              7     9.0

In [6]:
#looking at the columns of the orders_products_merged data set
ords_prods_merged.columns

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', '_merge', 'product_name', 'aisle_id',
       'department_id', 'prices'],
      dtype='object')

In [7]:
# a function that can devide columns values to 4 types depending on the value number ( here the price )
def price_label(row):
    if row['prices'] <= 5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High range'
    else: return 'Not enough data'

In [8]:
# assigning as part of the data set to variable for testing reasings
a=ords_prods_merged.head(100)

In [9]:
#trying with a small example with the set for testing reasons.
a['price_range'] = a.apply(price_label, axis=1)

C:\Users\Jalon\AppData\Local\Temp/ipykernel_17316/1789237346.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  a['price_range'] = a.apply(price_label, axis=1)


In [10]:
a

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
..       ...      ...      ...           ...                 ...   
95   3226575      360    prior             1                   5   
96   1469869      377    prior             3                   5   
97   1927023      387    prior             2                   4   
98    858092      420    prior             4                   1   
99    119314      420    prior             5                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
..                ...                     ...         ...                ...   
95                 12                     NaN         196                  1   
96                 17                     3.0         196                  9   
97                 10                    22.0         196                  3   
98                 19                    30.0         196                  2   
99                 19                    16.0         196                  2   

    reordered _merge product_name  aisle_id  department_id  prices  \
0           0   both         Soda        77              7     9.0   
1           1   both         Soda        77              7     9.0   
2           1   both         Soda        77              7     9.0   
3           1   both         Soda        77              7     9.0   
4           1   both         Soda        77              7     9.0   
..        ...    ...          ...       ...            ...     ...   
95          0   both         Soda        77              7     9.0   
96          0   both         Soda        77              7     9.0   
97          0   both         Soda        77              7     9.0   
98          0   both         Soda        77              7     9.0   
99          1   both         Soda        77              7     9.0   

          price_range  
0   Mid-range product  
1   Mid-range product  
2   Mid-range product  
3   Mid-range product  
4   Mid-range product  
..                ...  
95  Mid-range product  
96  Mid-range product  
97  Mid-range product  
98  Mid-range product  
99  Mid-range product  

[100 rows x 16 columns]

In [11]:
#applying the function pricce_label to the data set orders_products_merged and making a new column with the results
ords_prods_merged['price_range'] = ords_prods_merged.apply(price_label, axis=1)

In [12]:
ords_prods_merged.columns

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', '_merge', 'product_name', 'aisle_id',
       'department_id', 'prices', 'price_range'],
      dtype='object')

In [31]:
# preparing a list that devides the column orders_day_of_week to 3 different day types depending on the day's business
result = []

for value in ords_prods_merged["orders_day_of_week"]:
    if value == 0:
        result.append("Busiest day")
    elif value == 4:
        result.append("Least busy")
    else:
        result.append("Regularly busy")

In [33]:
#assignging the result list to a new column in the orders_products_merged data set
ords_prods_merged["busiest_day"] = result

In [34]:
ords_prods_merged.head(10)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   
5   3367565        1    prior             6                   2   
6    550135        1    prior             7                   1   
7   3108588        1    prior             8                   1   
8   2295261        1    prior             9                   1   
9   2550362        1    prior            10                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   
5                  7                    19.0         196                  1   
6                  9                    20.0         196                  1   
7                 14                    14.0         196                  2   
8                 16                     0.0         196                  4   
9                  8                    30.0         196                  1   

   reordered _merge product_name  aisle_id  department_id  prices  \
0          0   both         Soda        77              7     9.0   
1          1   both         Soda        77              7     9.0   
2          1   both         Soda        77              7     9.0   
3          1   both         Soda        77              7     9.0   
4          1   both         Soda        77              7     9.0   
5          1   both         Soda        77              7     9.0   
6          1   both         Soda        77              7     9.0   
7          1   both         Soda        77              7     9.0   
8          1   both         Soda        77              7     9.0   
9          1   both         Soda        77              7     9.0   

         price_range     busiest_day  
0  Mid-range product  Regularly busy  
1  Mid-range product  Regularly busy  
2  Mid-range product  Regularly busy  
3  Mid-range product      Least busy  
4  Mid-range product      Least busy  
5  Mid-range product  Regularly busy  
6  Mid-range product  Regularly busy  
7  Mid-range product  Regularly busy  
8  Mid-range product  Regularly busy  
9  Mid-range product      Least busy

In [13]:
ords_prods_merged["orders_day_of_week"]

0           2
1           3
2           3
3           4
4           4
           ..
32434207    2
32434208    5
32434209    2
32434210    3
32434211    4
Name: orders_day_of_week, Length: 32434212, dtype: int64

In [14]:
#checking the value_counts of the the orders_day_of_week column to prepare 
#a list depending on the frequency of the orders in those days
ords_prods_merged["orders_day_of_week"].value_counts()

0    6209632
1    5665830
6    4500246
2    4217766
5    4209449
3    3844096
4    3787193
Name: orders_day_of_week, dtype: int64

In [16]:
# preparing a list for the new column where the busiest days are aggregated
result2 = []

for value in ords_prods_merged["orders_day_of_week"]:
    if value == 0 or value ==1:
        result2.append("Busiest 2 days")
    elif value == 4 or value == 3:
        result2.append("Least busy 2 days")
    else:
        result2.append("Regularly busy")

In [17]:
# puting the finished list in the data set
ords_prods_merged["Busiest days"] = result2

In [19]:
# checking for the accuracy of the new column
ords_prods_merged["Busiest days"].value_counts()

Regularly busy       12927461
Busiest 2 days       11875462
Least busy 2 days     7631289
Name: Busiest days, dtype: int64

6209632 + 5665830 = 11875462 &
3844096 + 3787193 = 7631289 &
4500246 + 4217766 + 4209449 = 12927461
which means that the calculation of the new column are accurate

In [32]:
# checking the descriptive statistics of the distribution of the orders by hours of the day
ords_prods_merged["order_hour_of_day"].value_counts().describe()

count    2.400000e+01
mean     1.351426e+06
std      1.079354e+06
min      5.131700e+04
25%      2.728130e+05
50%      1.118200e+06
75%      2.558302e+06
max      2.764390e+06
Name: order_hour_of_day, dtype: float64

In [31]:
ords_prods_merged["order_hour_of_day"].value_counts().describe()

10    2764390
11    2738585
14    2691563
15    2664522
13    2663272
12    2620800
16    2537469
9     2456661
17    2089452
8     1719952
18    1637922
19    1259382
20     977017
7      891928
21     796362
22     634737
23     402612
6      290770
0      218942
1      115786
5       88057
2       69431
4       53283
3       51317
Name: order_hour_of_day, dtype: int64

mean: 1.351426e+06, 25%:  2.728130e+05, 50% : 1.118200e+06, 75%: 2.558302e+06

I am going to consider hours with value counts between 25% and 75 % as average and less than  2.728130e+05 (25%) will be considered as hours with fewest orders and value_counts with more than 2.558302e+06 (75%) will be considered as hours with most orders

In [41]:
#preparing the list for the new column with dividing the order hour of day to busiest period of day
#fewest orders and average orders depending on its value_counts
result3 = []

for value in ords_prods_merged["order_hour_of_day"]:
    if value in [10,11,12,13,14,15]:
        result3.append("busiest_period_of_day")
    elif value in [0,1,2,3,4,5]:
        result3.append("Fewest orders")
    else:
        result3.append("Average orders")

272813.0

In [43]:
#assignign the list to a new column in the order_prods_merged data set
ords_prods_merged["busiest_period_of_day"]= result3

In [47]:
# printing the frequency of the new column
print(ords_prods_merged["busiest_period_of_day"].value_counts())

busiest_period_of_day    16143132
Average orders           15694264
Fewest orders              596816
Name: busiest_period_of_day, dtype: int64


In [50]:
ords_prods_merged.isnull().sum()

order_id                        0
user_id                         0
eval_set                        0
order_number                    0
orders_day_of_week              0
order_hour_of_day               0
days_since_prior_order    2078084
product_id                      0
add_to_cart_order               0
reordered                       0
_merge                          0
product_name                28171
aisle_id                        0
department_id                   0
prices                          0
price_range                     0
Busiest days                    0
busiest_period_of_day           0
dtype: int64

As we can see the only missing values are in days_since_prior_order and product_name columns.
since we have product id, product name is for this task less important so i am going to keep it in case and for the days_since_prior_order as said in the last task this could just be the result of "first" orders so i am going to keep them as well

In [54]:
# exporting the ords_prods_merged data set
ords_prods_merged.to_pickle(os.path.join(path,
'ords_prods_merged.pkl'))